### PR TITLE
Feat/max file bytes env cap

### PR DESCRIPTION
**Pull request by SlayerBirden**
Allow deployments to bound in-memory Drive/Gmail/Chat downloads via an
env var (default unset/0 = uncapped) so large files cannot OOM the
process. Oversized downloads return an agent-friendly error with a Drive
link and alternative tools instead of buffering the full body.

## Description
Adds optional `WORKSPACE_MCP_MAX_FILE_BYTES` to cap how many bytes tools
may buffer into process memory when downloading Drive files, Gmail
attachments, Chat attachments, and Docs export/binary paths.

- **Default unchanged:** unset or `0` → uncapped (backwards compatible).
- **When set** (e.g. `5242880` for 5 MiB):
  - Pre-checks Drive `size` for binary downloads when metadata is available.
  - Streams the media response (httpx) and aborts on `Content-Length` or
    mid-body once buffered bytes exceed the limit — without relying on
    `MediaIoBaseDownload` / HTTP `Range`.
  - Returns an agent-friendly error with file metadata, Drive link, and
    suggested alternatives (`get_doc_content`, `read_sheet_values`, etc.)
    instead of loading the full body (which is especially wasteful in
    stateless mode, where only a base64 preview is returned anyway).

**Why stream instead of `MediaIoBaseDownload` when capped:** Drive
`export_media` (e.g. Google Docs → PDF) often ignores `Range` and returns
HTTP 200 with the entire body. `MediaIoBaseDownload` buffers that full
`content` before any size check can run, so a chunksize alone cannot
enforce the cap for exports.

**Uncapped path:** still uses `MediaIoBaseDownload` and returns the same
full download result. Chunk size is set to **256 KiB** instead of the
library default (**100 MiB**) so a single `next_chunk()` cannot pin ~100 MiB
even when no env cap is set. Same bytes returned; only transport chunking
changes (more smaller Range requests when Drive honors Range). Not a
behavioral break for callers.

Shared helpers live in `core/file_limits.py` and are wired into Drive,
Gmail, Docs, and Chat download paths.

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## Testing
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have tested this change manually

Added:
- `tests/core/test_file_limits.py` — env parsing, pre-check, Content-Length reject, mid-stream abort
- `tests/gdrive/test_file_size_limit.py` — tool-level Drive reject / allow paths

Locally: `uv run pytest tests/core/test_file_limits.py tests/gdrive/test_file_size_limit.py tests/gdrive/test_drive_file_content.py`

Manual: with `WORKSPACE_MCP_MAX_FILE_BYTES=1000`, `get_drive_file_download_url` on a Google Doc export rejects after ~2 KiB streamed (not after buffering the full ~277 KiB PDF).

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] **I have enabled "Allow edits from maintainers" for this pull request**

## Additional Notes
Motivated by production ECS/Fargate OOM kills where
`get_drive_file_download_url` buffered ~97 MiB files into memory under
`WORKSPACE_MCP_STATELESS_MODE` (full download, stub response). This cap is
opt-in so self-hosted / stdio users who rely on large local downloads are
unaffected unless they set the env var.

Suggested production starting point: `WORKSPACE_MCP_MAX_FILE_BYTES=5242880`
(or `10485760`), tuned to task memory and traffic.

---

**⚠️ IMPORTANT:** This repository requires that you enable "Allow edits from maintainers" when creating your pull request. This allows maintainers to make small fixes and improvements directly to your branch, speeding up the review process.

To enable this setting:
1. When creating the PR, check the "Allow edits from maintainers" checkbox
2. If you've already created the PR, you can enable this in the PR sidebar under "Allow edits from maintainers"

## Release Notes

* **New Features**
  * Added configurable file-size limits for Google Drive, Gmail, Chat, and Docs downloads.
  * Configure limits with `WORKSPACE_MCP_MAX_FILE_BYTES`; downloads remain uncapped when unset or set to `0`.
  * Oversized in-memory downloads are rejected with clear error messages before or during download.
  * Downloads streamed directly to disk are not subject to the configured limit.

* **Bug Fixes**
  * Improved reliability when updating Drive files concurrently.

* **Documentation**
  * Clarified file-size limit behavior, validation, and supported download types.

### CODE REVIEW

File: tests/core/test_file_limits.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    client_cm.__aenter__.return_value = client
+    client_cm.__aexit__.return_value = None
+    return client_cm, resp
+
+
+@pytest.mark.asyncio
+async def test_download_media_bytes_rejects_via_content_length(monkeypatch):
+    """Drive export often sends Content-Length and ignores Range — reject before body."""
+    monkeypatch.setenv("WORKSPACE_MCP_MAX_FILE_BYTES", "1000")
+    request = Mock(uri="https://www.googleapis.com/drive/v3/files/x/export", method="GET")
```
[the remark this answers is not in the document]
Comment (SlayerBirden) [reply tests/core/test_file_limits.py]: already fixed

File: core/file_limits.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+                # googleapiclient.HttpError expects an httplib2-style resp.status.
+                httplib2_resp = type(
+                    "Response",
+                    (),
+                    {
+                        "status": resp.status_code,
+                        "reason": resp.reason_phrase,
+                    },
+                )()
+                raise HttpError(httplib2_resp, body, uri=uri)
```
[the remark this answers is not in the document]
Comment (SlayerBirden) [reply core/file_limits.py]: Acknowledged — thanks for flagging this.

We're **not addressing it in this PR** on purpose.

**What's valid:** with `follow_redirects=True`, httpx does `aread()` on intermediate redirect responses, and our ≥400 path also uses unbounded `aread()`. In theory those bodies could bypass `WORKSPACE_MCP_MAX_FILE_BYTES`.

**Why we're deferring:**
1. **Not the production failure mode.** OOMs came from successful large media downloads (e.g. ~97 MiB Drive `get_media` / exports under stateless mode), not from redirect or error payloads.
2. **Practical risk is low for these Google endpoints.** Drive/Chat media redirects are typically empty/tiny 3xx + `Location`; Google API error bodies are small JSON, not multi‑MB blobs.
3. **Scope / cost.** Properly fixing this means disabling auto-redirects, manually following `Location` while applying the byte cap to every hop, bounding error-body reads, and adding regression tests — real complexity for a defense-in-depth edge case outside this PR's goal (cap successful download buffering).

Happy to take it as a follow-up if maintainers want belt-and-suspenders hardening; it shouldn't block the opt-in download cap.

**Comment by taylorwilsdon** [comment]
This one took some massaging but I think we're there, thanks!
